[RFC] Make `zephyr,memory-attr` a capabilities bitmask

carlocaione opened the pull request:
This is the final step in making the `zephyr,memory-attr` property actually useful.

The problem with the current implementation is that `zephyr,memory-attr` is an enum type, this is making very difficult to use that to actually describe the memory capabilities. The solution proposed in this PR is to use the `zephyr,memory-attr` property as an OR-ed bitmask of memory attributes.

> [!IMPORTANT]  
> With the change proposed in this PR it is possible in the DeviceTree to mark the memory regions with a bitmask of attributes by using the `zephyr,memory-attr` property. This property and the related memory region can then be retrieved at run-time by leveraging a provided helper library or the usual DT helpers.

The set of general attributes that can be specified in the property are defined and explained in `include/zephyr/dt-bindings/memory-attr/memory-attr.h` (the list can be extended when needed).

https://github.com/zephyrproject-rtos/zephyr/blob/5664796d1f5b525dc27bb6dfc862c35e7c5b7b13/include/zephyr/dt-bindings/memory-attr/memory-attr.h#L16-L20

For example, to mark a memory region in the DeviceTree as volatile, non-cacheable, out-of-order:

```
   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-attr = <( DT_MEM_VOLATILE | DT_MEM_NON_CACHEABLE | DT_MEM_OOO )>;
   };
```

> [!IMPORTANT]  
> The `zephyr,memory-attr` property can also be used to set architecture-specific custom attributes that can be interpreted at run time. This is leveraged, among other things, to create MPU regions out of DeviceTree defined memory regions on ARM, for example:

```
   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-region = "NOCACHE_REGION";
       zephyr,memory-attr = <( DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE) )>;
   };
```
See  `include/zephyr/dt-bindings/memory-attr/memory-attr-mpu.h` to see how an architecture can define its own special memory attributes (in this case ARM MPU).

> [!IMPORTANT]  
> The property can also be used to set custom software-specific attributes. For example we can think of marking a memory region as available to be used for memory allocation (not yet implemented):

```
   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-attr = <( DT_MEM_NON_CACHEABLE | DT_MEM_SW_ALLOCATABLE )>;
   };
```

Or maybe we can leverage the property to specify some alignment requirements for the region:
```
   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-attr = <( DT_MEM_CACHEABLE | DT_MEM_SW_ALIGN(32) )>;
   };
```

> [!IMPORTANT]  
> The conventional and recommended way to deal and manage with memory regions marked with attributes is by using the provided `mem-attr` helper library by enabling `CONFIG_MEM_ATTR` (or by using the usual DT helpers).

When this option is enabled the list of memory regions and their attributes are compiled in a user-accessible array and a set of functions is made available that can be used to query, probe and act on regions and attributes, see `include/zephyr/mem_mgmt/mem_attr.h`

> [!NOTE] 
> Note that the `zephyr,memory-attr` property is only a descriptive property of the capabilities of the associated memory  region, but it does not result in any actual setting for the memory to be set. The user, code or subsystem willing to use this information to do some work (for example creating an MPU region out of the property) must use either the provided `mem-attr` library or the usual DeviceTree helpers to perform the required work / setting.

### Note: how to review this PR
>  [!WARNING]  
> Unfortunately this PR cannot be divided into smaller PRs to not break bisectability.

In general this PR we are:
- Removing the previous DT-based helpers for `zephyr,memory-attr`, these were recently introduced but proved to not be up to the task
- Adding the `int` version of the DT property, see `dts/bindings/base/zephyr,memory-attr.yaml`, `include/zephyr/dt-bindings/memory-attr/memory-attr.h` and `include/zephyr/dt-bindings/memory-attr/memory-attr-mpu.h`
- Adding the `mem-attr` library, see `include/zephyr/mem_mgmt/mem_attr.h` and `subsys/mem_mgmt/mem_attr.c`
- Reworking the ARM MPU drivers to use the `mem-attr` to create the DT-defined MPU regions: see `arch/arm/core/aarch32/mpu/arm_mpu.c` and `arch/arm64/core/cortex_r/arm_mpu.c`
- Fixing all the DTS and moving from `enum` to `int`
- Adding documentation in `doc/services/mem_mgmt/index.rst`

carlocaione commented:
CI failure is unrelated.

hubertmis commented:
I'm a little confused. I understand I can use `zephyr,memory-attr = <( DT_MEM_VOLATILE | DT_MEM_NON_CACHEABLE | DT_MEM_OOO )>;` or `zephyr,memory-attr = <( DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE) )>;`.

But would these result in the same MPU entry if added to an ARM's .dts file?

carlocaione commented:
> I'm a little confused. I understand I can use `zephyr,memory-attr = <( DT_MEM_VOLATILE | DT_MEM_NON_CACHEABLE | DT_MEM_OOO )>;` or `zephyr,memory-attr = <( DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE) )>;`.
> 
> But would these result in the same MPU entry if added to an ARM's .dts file?

No, you can create an `MPU` entry only using the `DT_ARM_MPU(...)` macro and that is to be considered legacy and it is there only to support the current way we have in Zephyr to create MPU regions.

`zephyr,memory-attr = <( DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE) )>;` is creating an MPU region because in the ARM MPU driver I added the code to do that to be backward compatible with the current way we have to create MPU regions.

`zephyr,memory-attr = <( DT_MEM_VOLATILE | DT_MEM_NON_CACHEABLE | DT_MEM_OOO )>;` is only describing the kind of memory. it COULD be used to create an MPU regions but currently there is no code doing that.

I tried to be as much clear as possible here: https://github.com/zephyrproject-rtos/zephyr/blob/5664796d1f5b525dc27bb6dfc862c35e7c5b7b13/doc/services/mem_mgmt/index.rst?plain=1#L57-L64

and here:
https://github.com/zephyrproject-rtos/zephyr/blob/5664796d1f5b525dc27bb6dfc862c35e7c5b7b13/include/zephyr/dt-bindings/memory-attr/memory-attr-mpu.h#L12-L20

hubertmis commented:
> No, you can create an `MPU` entry only using the `DT_ARM_MPU(...)` macro and that is to be considered legacy and it is there only to support the current way we have in Zephyr to create MPU regions.

Thanks. Now I understand how these attributes relate.
So what would be a proper way to describe these memory attributes and create an associated MPU entry? Something like this?

```
zephyr,memory-attr = <( DT_MEM_VOLATILE | DT_MEM_NON_CACHEABLE | DT_MEM_OOO | DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE) )>;
```

carlocaione commented:
> Thanks. Now I understand how these attributes relate. So what would be a proper way to describe these memory attributes and create an associated MPU entry? Something like this?
> 
> ```
> zephyr,memory-attr = <( DT_MEM_VOLATILE | DT_MEM_NON_CACHEABLE | DT_MEM_OOO | DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE) )>;
> ```

Correct. While this can be much extended we can also do some kind of checking on the correct attributes when creating an MPU region like I did here to verify that we do not set `DT_MEM_CACHEABLE` with `DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE)`: https://github.com/zephyrproject-rtos/zephyr/blob/5664796d1f5b525dc27bb6dfc862c35e7c5b7b13/arch/arm/core/aarch32/mpu/arm_mpu.c#L103-L107

hubertmis commented:
Sounds good. I think it would be perfect if we could deprecate arch-specific attributes like `DT_ARM_MPU()` and create a memory configuration based only on the generic attributes.

But we can discuss this in the future. This RFC looks like a step in a good direction.

carlescufi commented:
Architecture WG:

- @galak asks if this breaks the `zephyr,memory-attr` backwards compat, @carlocaione mentions that he has shown how to transition in the PR
- @teburd asks if TCM is included in the memory types, @carlocaione says that this could be extended (32 bits)

galak reviewed (dismissed):
Want to make sure we get input/mapping for multiple architectures for this (arm, arc, xtensa, risc-v, x86).

teburd commented:
Big +1 from me.

I think setting some rules/guidelines around what these bits/attributes/capabilities should be used for would be good. The bit flags in particular I think are nice as it removes the quirky additive enum naming that was here before.

carlocaione commented:
> Want to make sure we get input/mapping for multiple architectures for this (arm, arc, xtensa, risc-v, x86).

@galak the idea is that we have a set of attributes that are generic: https://github.com/zephyrproject-rtos/zephyr/blob/5664796d1f5b525dc27bb6dfc862c35e7c5b7b13/include/zephyr/dt-bindings/memory-attr/memory-attr.h#L11-L21

and then a mask is set for architecture-specific attributes so that each architecture can add its own binding file with custom attributes: https://github.com/zephyrproject-rtos/zephyr/blob/5664796d1f5b525dc27bb6dfc862c35e7c5b7b13/include/zephyr/dt-bindings/memory-attr/memory-attr.h#L23-L32


At this stage I thought that we could only agree on the common / generic set of attributes and then add the arch-dependent ones when needed, or do you want to have those as well?

carlocaione commented:
Opened #61102 to gather all the desiderata memory attributes from the various arch maintainers.

mbolivar-ampere commented:
@carlocaione did you consider deprecating `zephyr,memory-attr` and adding a new property (named something like `zephyr,mem-attr`) instead? Is that impractical?

povergoing commented on arch/arm64/core/cortex_r/arm_mpu.c:
```diff
@@ -191,6 +252,12 @@ void z_arm64_mm_init(bool is_primary_core)
+		__ASSERT(0, "Failed to allocate MPU regions from DT\n");
+		return;
+	}
+
```
Hi Carlo, I saw there is a difference compared to ARM. This should be put below the following line `static_regions_num = mpu_config.num_regions;`

```diff
diff --git a/arch/arm64/core/cortex_r/arm_mpu.c b/arch/arm64/core/cortex_r/arm_mpu.c
index f1486bc445..c0b1724057 100644
--- a/arch/arm64/core/cortex_r/arm_mpu.c
+++ b/arch/arm64/core/cortex_r/arm_mpu.c
@@ -252,15 +252,15 @@ void z_arm64_mm_init(bool is_primary_core)
                region_init(r_index, &mpu_config.mpu_regions[r_index]);
        }

+       /* Update the number of programmed MPU regions. */
+       static_regions_num = mpu_config.num_regions;
+
        /* DT-defined MPU regions. */
        if (mpu_configure_regions_from_dt(&static_regions_num) == -EINVAL) {
                __ASSERT(0, "Failed to allocate MPU regions from DT\n");
                return;
        }

-       /* Update the number of programmed MPU regions. */
-       static_regions_num = mpu_config.num_regions;
-
        arm_core_mpu_enable();

        if (!is_primary_core) {
```

carlocaione replied:
well spotted, thanks! I'll fix that after gathering more comments and reviews.

carlocaione commented:
> @carlocaione did you consider deprecating `zephyr,memory-attr` and adding a new property (named something like `zephyr,mem-attr`) instead? Is that impractical?

@mbolivar-ampere define impractical :) that would require a bit of work because there would be overlapping between the two properties when generating the MPU regions, so we should exclude one when also the other one is present.

Now, that said, I have to shamefully come forward and clean and say that this doesn't really matter because I already forgot to do the due diligence in #60049 about deprecation. So since I cannot change the past, I'll do the whole deprecation dance here with this PR so that we can deprecate the old already-removed ENUM `zephyr,memory-region-mpu` for the INT `zephy,memory-attr`.

On the good side the old `zephyr,memory-region-mpu` was ARM+MPU specific so not widely used.

mbolivar-ampere commented:
> so we should exclude one when also the other one is present.

Makes sense -- to keep things simple, though, you could maybe just BUILD_ASSERT that only one is present?

mbolivar-ampere commented:
Ah, OK, thanks for clarifying. I agree that there's no need to deprecate something that has never been part of a release.

carlocaione committed:
dt: Make zephyr,memory-attr a capabilities bitmask

This is the final step in making the `zephyr,memory-attr` property
actually useful.

The problem with the current implementation is that `zephyr,memory-attr`
is an enum type, this is making very difficult to use that to actually
describe the memory capabilities. The solution proposed in this PR is to
use the `zephyr,memory-attr` property as an OR-ed bitmask of memory
attributes.

With the change proposed in this PR it is possible in the DeviceTree to
mark the memory regions with a bitmask of attributes by using the
`zephyr,memory-attr` property. This property and the related memory
region can then be retrieved at run-time by leveraging a provided helper
library or the usual DT helpers.

The set of general attributes that can be specified in the property are
defined and explained in
`include/zephyr/dt-bindings/memory-attr/memory-attr.h` (the list can be
extended when needed).

For example, to mark a memory region in the DeviceTree as volatile,
non-cacheable, out-of-order:

   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-attr = <( DT_MEM_VOLATILE |
			       DT_MEM_NON_CACHEABLE |
			       DT_MEM_OOO )>;
   };

The `zephyr,memory-attr` property can also be used to set
architecture-specific custom attributes that can be interpreted at run
time. This is leveraged, among other things, to create MPU regions out
of DeviceTree defined memory regions on ARM, for example:

   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-region = "NOCACHE_REGION";
       zephyr,memory-attr = <( DT_ARM_MPU(ATTR_MPU_RAM_NOCACHE) )>;
   };

See `include/zephyr/dt-bindings/memory-attr/memory-attr-mpu.h` to see
how an architecture can define its own special memory attributes (in
this case ARM MPU).

The property can also be used to set custom software-specific
attributes. For example we can think of marking a memory region as
available to be used for memory allocation (not yet implemented):

   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-attr = <( DT_MEM_NON_CACHEABLE |
			       DT_MEM_SW_ALLOCATABLE )>;
   };

Or maybe we can leverage the property to specify some alignment
requirements for the region:

   mem: memory@10000000 {
       compatible = "mmio-sram";
       reg = <0x10000000 0x1000>;
       zephyr,memory-attr = <( DT_MEM_CACHEABLE |
			       DT_MEM_SW_ALIGN(32) )>;
   };

The conventional and recommended way to deal and manage with memory
regions marked with attributes is by using the provided `mem-attr`
helper library by enabling `CONFIG_MEM_ATTR` (or by using the usual DT
helpers).

When this option is enabled the list of memory regions and their
attributes are compiled in a user-accessible array and a set of
functions is made available that can be used to query, probe and act on
regions and attributes, see `include/zephyr/mem_mgmt/mem_attr.h`

Note that the `zephyr,memory-attr` property is only a descriptive
property of the capabilities of the associated memory  region, but it
does not result in any actual setting for the memory to be set. The
user, code or subsystem willing to use this information to do some work
(for example creating an MPU region out of the property) must use either
the provided `mem-attr` library or the usual DeviceTree helpers to
perform the required work / setting.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

erwango commented on boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:
```diff
@@ -50,7 +51,7 @@
 		reg = <0xd0000000 DT_SIZE_M(16)>;
 		zephyr,memory-region = "SDRAM2";
-		zephyr,memory-attr = "RAM";
+		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_RAM) )>;
```
Any reason to add parenthesis ?

Following would be lighter:
```suggestion
		zephyr,memory-attr = <DT_MEM_ARM(ATTR_MPU_RAM)>;
```

carlocaione replied:
That's actually required otherwise otherwise DTC will complain with: `parse error: expected number or parenthesized expression`

dcpleung commented:
LGTM on Xtensa.

mbolivar-ampere commented on arch/arm/core/aarch32/mpu/arm_mpu.c:
```diff
@@ -32,6 +34,13 @@ LOG_MODULE_DECLARE(mpu);
 #define NUM_MPU_REGIONS   DT_PROP(MPU_NODEID, arm_num_mpu_regions)
 #endif
 
+#define DT_NODE_HAS_PROP_AND_OR(node_id, prop) \
```
This is use of the reserved DT namespace and needs to be renamed; I suggest NODE_HAS_PROP_OR or something like that.

carlocaione replied:
Fair enough, I'll rename it.

carlocaione replied:
done

mbolivar-ampere commented on doc/releases/release-notes-3.5.rst:
```diff
@@ -35,6 +35,17 @@ Changes in this release
+  moved from 'enum' to 'int'. To have a seamless conversion this is the
+  required change in the DT:
+
+    - "RAM"         -> <( DT_MEM_ARM(ATTR_MPU_RAM) )>
```
Does this look OK in the html? I noticed that below this text is put in a code block, but that's not present here.

carlocaione replied:
does the release note rendered in html? In that case yes, the code block is missing.

carlocaione replied:
done

mbolivar-ampere reviewed (commented):
Thanks for the nice doc and for a smooth migration path. I have a few minor questions and nitpicks

carlocaione commented on dts/bindings/base/zephyr,memory-attr.yaml:
```diff
@@ -4,17 +4,19 @@
+    deprecated: true
+    description: |
+      Signify that this node should result in a dedicated MPU region.
+      Deprecated in favor of `zephyr,memory-attr`.
```
it's just to emphasize the property and make it stand out a bit but I can change it.

carlocaione replied:
changed with `'`/

carlocaione commented on include/zephyr/mem_mgmt/mem_attr.h:
```diff
@@ -0,0 +1,114 @@
+
+/** @cond INTERNAL_HIDDEN */
+
+#define _DT_MEM_ATTR	zephyr_memory_attr
```
Yes there is a reason. I have always used names beginning with underscore to mark something that is reserved for the current file and usually not exported outside, so this is a perfect example of this use case (the underscore is basically saying "there is nothing to see, move along"). It is also not inlined because I want this to stands out when reading the header file for self-documentation.

carlocaione replied:
Changed with `__MEM_ATTR`

carlocaione commented on include/zephyr/mem_mgmt/mem_attr.h:
```diff
@@ -0,0 +1,114 @@
+ * @param fn macro to invoke
+ */
+#define DT_MEMORY_ATTR_FOREACH_STATUS_OKAY_NODE(fn)			\
+	DT_FOREACH_STATUS_OKAY_NODE_VARGS(_FILTER, fn)
```
not particularly opinionated about this. I'll try without the `VARGS` variant.

carlocaione replied:
Ok, I thought that this was feasible but now I'm a bit lost. The point here is that I want to use that macro as:
```c
#define WHATEVER(node_id) ...

static const struct here[] = {
    DT_MEMORY_ATTR_FOREACH_STATUS_OKAY_NODE(WHATEVER)
}
```
so it looks to me that I need to use `DT_FOREACH_STATUS_OKAY_NODE_VARGS` here? Or am I missing your point?

carlocaione commented on subsys/mem_mgmt/Kconfig:
```diff
@@ -0,0 +1,12 @@
+
+config MEM_ATTR
+	bool "Memory Attributes management library"
+	default y if ARM_MPU
```
When the MPU is used we must enable this. The MPU drivers are using the mem attr library to build the regions mapping at run time instead of building that at compile time.

About the size increase I would not be concerned, we are moving the mappings from compile-time to run-time but yes, the library is slightly increasing the code size.

carlocaione commented on include/zephyr/mem_mgmt/mem_attr.h:
```diff
@@ -0,0 +1,114 @@
+ */
+struct mem_attr_region_t {
+	/** Memory node full name */
+	const char *dt_name;
```
It is, or future use.

carlocaione commented on doc/services/mem_mgmt/index.rst:
```diff
@@ -0,0 +1,90 @@
+library.
+
+The set of general attributes that can be specified in the property are defined
+and explained in :file:`include/zephyr/dt-bindings/memory-attr/memory-attr.h`.
```
interesting, I'll do.

carlocaione replied:
done

carlocaione commented on arch/arm64/core/cortex_r/arm_mpu.c:
```diff
@@ -143,6 +152,64 @@ static void region_init(const uint32_t index,
+				}
+
+/* This internal function programs the MPU regions defined in the DT when using
+ * the `zephyr,memory-attr = <( DT_MEM_ARM(...) >)` property.
```
done

carlescufi commented:
> Want to make sure we get input/mapping for multiple architectures for this (arm, arc, xtensa, risc-v, x86).

Given that @teburd and @evgeniy-paltsev are both in the conversation I will dismiss the -1 after we get a +1 from both.

dcpleung commented on include/zephyr/dt-bindings/memory-attr/memory-attr-arm.h:
```diff
@@ -0,0 +1,40 @@
+#define  ATTR_MPU_PPB			BIT(3)
+#define  ATTR_MPU_IO			BIT(4)
+#define  ATTR_MPU_EXTMEM		BIT(5)
+#define  ATTR_MPU_UNKNOWN		BIT(11) /* must be last */
```
I wonder if all these `ATTR_<arch>_UNKNOWN` can be alias to something like `DT_MEM_ARCH_UNKNOWN`. If we ever need to change the allocation of bits, we don't have to change all of these `BIT(11)`.

carlocaione replied:
yes, makes sense. I'll make the change.

povergoing reviewed (dismissed):
refresh +1

carlescufi commented:
@carlocaione please rebase

mbolivar-ampere commented on doc/services/mem_mgmt/index.rst:
```diff
@@ -0,0 +1,90 @@
+Memory Attributes
+#################
+
+It is possible in the DeviceTree to mark the memory regions with attributes by
```
nonblocking: 'devicetree' (or 'Devicetree' at beginning of sentence) is what we do, for consistency with the DT spec. This applies elsewhere in the patch too

mbolivar-ampere commented on doc/services/mem_mgmt/index.rst:
```diff
@@ -0,0 +1,90 @@
+   out of the property) must use either the provided ``mem-attr`` library or
+   the usual DeviceTree helpers to perform the required work / setting.
+
+A sample/test for the ``mem-attr`` library and its usage is provided in
```
Seems a bit of a stretch to call this a sample

carlocaione replied:
Yeah, changed that.

mbolivar-ampere commented on doc/services/mem_mgmt/index.rst:
```diff
@@ -0,0 +1,90 @@
+When the ``zephyr,memory-attr`` property was introduced, the
+``zephyr,memory-region-mpu`` property was removed and deprecated.
+
+The developers that are still using the deprecated property can move to the new
```
Thanks for this!

carlocaione replied:
❤️

mbolivar-ampere reviewed (commented):
LGTM, a couple of nonblocking suggestions. Thanks for taking the time to document the transition guide.

carlocaione commented:
Adding DNM because this needs to be squashed before merging.

carlocaione commented:
@teburd @erwango @povergoing @mbolivar-ampere @dcpleung sorry to ask you to ACK this one more time but I had to squash all the commits for the sake of bisectability.
